### PR TITLE
Implement MSC3952: Intentional mentions

### PR DESCRIPTION
**Pull request by clokep**
MSC3952 defines push rules which act on a list of MXIDs in the event body, instead of searching the entire event body for display name / local part.

This implements the above in Synapse behind an experimental configuration flag. It does not yet implement the backwards compatibility pieces of the MSC.

### CODE REVIEW

**Comment by clokep** [comment]
There's still a couple of unknowns on the MSC which might affect:

* The exact format of the field (including if the length is limited).
* Backwards compatibility.

I think since this is disabled by default and subject to change that this is likely ready for an initial review and we can go from there.

**Comment by clokep** [comment]
(Also -- I have other experimental push changes and I don't want to conflict. 👍 )

File: rust/src/push/mod.rs
Comment (DMRobertson) [inline rust/src/push/mod.rs:546]: Would it be worth tests that the `Condition` instance also serialises correctly to that piece of json? (It doesn't seem to be the done thing in this file, so probably not?)

File: synapse/push/bulk_push_rule_evaluator.py
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py]: Does the MSC impose a limit on the number of mentions?

We're ignoring the 11th mention and onwards here, rather than rejecting the event outright. I suppose that's probably a lot more practical, because otherwise you'd have to update the auth rules and define a new room version.

File: stubs/synapse/synapse_rust/push.pyi
Comment (DMRobertson) [inline stubs/synapse/synapse_rust/push.pyi:59]: Would `AbstractSet` make sense here, to permit passing both frozensets and sets to this function?

File: tests/push/test_bulk_push_rule_evaluator.py
Comment (DMRobertson) [inline tests/push/test_bulk_push_rule_evaluator.py:140]: Took me a moment to realise what this function does. Maybe a docstring e.g. 

```suggestion
            """Returns true iff the `mentions` trigger an event push action."""
            content = {}
```

would help?

**Review by DMRobertson — COMMENTED** [review]
Generally LGTM :+1:

IIRC the ability to send `@room` notifications --- or rather `{"m.notify": {"room": true}}` notifications --- can be (is?) gated based the sender's power level. Is that still the case after the MSC? I ask because I couldn't see anything about power levels here. I'm guessing Synapse enforces that when processing the `event_push_actions_staging` table?

File: stubs/synapse/synapse_rust/push.pyi
Comment (clokep) [reply stubs/synapse/synapse_rust/push.pyi:59]: It doesn't work properly with pyo3 when I tested.

File: synapse/push/bulk_push_rule_evaluator.py
Comment (clokep) [reply synapse/push/bulk_push_rule_evaluator.py]: This was removed from the PR.

**Comment by clokep** [comment]
> IIRC the ability to send `@room` notifications --- or rather `{"m.notify": {"room": true}}` notifications --- can be (is?) gated based the sender's power level. Is that still the case after the MSC? I ask because I couldn't see anything about power levels here. I'm guessing Synapse enforces that when processing the `event_push_actions_staging` table?

Yes, I failed to include the extra check in the push rule definition though -- it is handled by a separate condition, (`sender_notification_permission`). Will update the PR. (Note that MSC is correct here.)

File: synapse/push/bulk_push_rule_evaluator.py
Comment (DMRobertson) [reply synapse/push/bulk_push_rule_evaluator.py]: So it was. But to check, there's no limit on the number of mentions now?

There's a mention under ["abuse potential"](https://github.com/matrix-org/matrix-spec-proposals/blob/clokep/intentional-mentions/proposals/3952-intentional-mentions.md#abuse-potential) that moderation tooling can limit the number of mentions. I didn't follow that---what's the thinking? Is the `m.mentions` field redactable or replaceable somehow?

Comment (clokep) [reply synapse/push/bulk_push_rule_evaluator.py]: > There's a mention under ["abuse potential"](https://github.com/matrix-org/matrix-spec-proposals/blob/clokep/intentional-mentions/proposals/3952-intentional-mentions.md?rgh-link-date=2023-01-27T00%3A36%3A06Z#abuse-potential) that moderation tooling can limit the number of mentions. I didn't follow that---what's the thinking? Is the `m.mentions` field redactable or replaceable somehow?

You use something like mjolnir to ban/kick people who are abusive or you use a Synapse module to reject events with too many mentions.

File: stubs/synapse/synapse_rust/push.pyi
Comment (clokep) [reply stubs/synapse/synapse_rust/push.pyi:59]: For more info, if I default to using a `frozenset()` if e.g. the `m.mentions` field doesn't exist, then we throw this error when attempting to pass that to Rust:

```
	Traceback (most recent call last):
	  File "/Users/clokep/matrix/synapse/synapse/http/server.py", line 307, in _async_render_wrapper
	    callback_return = await self._async_render(request)
	  File "/Users/clokep/matrix/synapse/synapse/http/server.py", line 513, in _async_render
	    callback_return = await raw_callback_return
	  File "/Users/clokep/matrix/synapse/synapse/rest/client/room.py", line 162, in on_POST
	    info, _ = await self._room_creation_handler.create_room(
	  File "/Users/clokep/matrix/synapse/synapse/handlers/room.py", line 926, in create_room
	    ) = await self._send_events_for_new_room(
	  File "/Users/clokep/matrix/synapse/synapse/handlers/room.py", line 1148, in _send_events_for_new_room
	    ev = await self.event_creation_handler.handle_new_client_event(
	  File "/Users/clokep/matrix/synapse/synapse/util/metrics.py", line 113, in measured_func
	    r = await func(self, *args, **kwargs)
	  File "/Users/clokep/matrix/synapse/synapse/handlers/message.py", line 1433, in handle_new_client_event
	    result, _ = await make_deferred_yieldable(
	  File "/Users/clokep/.pyenv/versions/synapse/lib/python3.9/site-packages/twisted/internet/defer.py", line 1697, in _inlineCallbacks
	    result = context.run(gen.send, result)
	  File "/Users/clokep/matrix/synapse/synapse/handlers/message.py", line 1469, in _persist_events
	    await self._bulk_push_rule_evaluator.action_for_events_by_user(
	  File "/Users/clokep/matrix/synapse/synapse/push/bulk_push_rule_evaluator.py", line 287, in action_for_events_by_user
	    await self._action_for_event_by_user(event, context, event_id_to_event)
	  File "/Users/clokep/matrix/synapse/synapse/util/metrics.py", line 113, in measured_func
	    r = await func(self, *args, **kwargs)
	  File "/Users/clokep/matrix/synapse/synapse/push/bulk_push_rule_evaluator.py", line 366, in _action_for_event_by_user
	    evaluator = PushRuleEvaluator(
	TypeError: argument 'user_mentions': 'frozenset' object cannot be converted to 'PySet'
```

Note that we could *always* use `frozenset` if we wanted, but I also wasn't really able to get that to work properly with pyo3.

File: synapse/push/bulk_push_rule_evaluator.py
Comment (DMRobertson) [reply synapse/push/bulk_push_rule_evaluator.py]: Thanks. I'm a little surprised---I'd bias towards having some kind of upper limit in the spec---but I'll defer to whatever the MSC process decides.

File: rust/src/push/mod.rs
Comment (clokep) [reply rust/src/push/mod.rs:546]: Maybe? I kind of just copied the other tests around here which seem to do this.

Comment (DMRobertson) [reply rust/src/push/mod.rs:546]: Ahh, okay. I interpreted it as "does serde do what I think it's going to do?". (I have sometimes written tests to that effect when playing with Pydantic...(

**Review by DMRobertson — APPROVED** [review]
LGTM, I'm happy if CI is!

**Pull request merged**
